drivers/tsl2561/tsl2561.c

```diff
@@ -0,0 +1,296 @@
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_tsl2561
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the TSL2561 Luminosity sensor.
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <math.h>
+#include <string.h>
+
+#include "log.h"
+#include "tsl2561.h"
+#include "tsl2561_internals.h"
+#include "periph/i2c.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+/* internal helpers */
+static void _enable(tsl2561_t *dev);
+static void _disable(tsl2561_t *dev);
+static void _read_data(tsl2561_t *dev, uint16_t *full, uint16_t *ir);
+
+#if ENABLE_DEBUG
+static void _print_init_info(tsl2561_t *dev)
+{
+    DEBUG("[Info] I2C device: %d\n", dev->i2c_dev);
+    DEBUG("[Info] Address: %d\n", dev->addr);
+    switch(dev->gain) {
+    case TSL2561_GAIN_1X:
+        DEBUG("[Info] Gain: 1X\n");
+        break;
+    case TSL2561_GAIN_16X:
+        DEBUG("[Info] Gain: 16X\n");
+        break;
+    default:
+        DEBUG("[Info] Invalid gain %d\n", dev->gain);
+        break;
+    }
+
+    switch(dev->integration) {
+    case TSL2561_INTEGRATIONTIME_13MS:
+        DEBUG("[Info] Integration time: 13ms\n");
+        break;
+    case TSL2561_INTEGRATIONTIME_101MS:
+        DEBUG("[Info] Integration time: 101ms\n");
+        break;
+    case TSL2561_INTEGRATIONTIME_402MS:
+        DEBUG("[Info] Integration time: 402ms\n");
+        break;
+    case TSL2561_INTEGRATIONTIME_NA:
+        DEBUG("[Info] Integration time: n/a\n");
+        break;
+    default:
+        DEBUG("[Info] Invalid integration time %d\n", dev->integration);
+        break;
+    }
+}
+#endif
+
+/*---------------------------------------------------------------------------*
+ *                          TSL2561 Core API                                 *
+ *---------------------------------------------------------------------------*/
+
+int tsl2561_init(tsl2561_t *dev,
+                 i2c_t i2c, uint8_t addr, uint8_t gain, uint8_t integration)
+{
+    dev->i2c_dev = i2c;
+    dev->addr = addr;
+    dev->gain = gain;
+    dev->integration = integration;
+
+#if ENABLE_DEBUG
+    _print_init_info(dev);
+#endif
+
+    /* Initialize I2C interface */
+    if (i2c_init_master(dev->i2c_dev, I2C_SPEED_NORMAL)) {
+        DEBUG("[Error] I2C device not enabled\n");
+        return -1;
+    }
+
+    DEBUG("[Info] I2C device initialized with success!\n");
+
+    /* Acquire exclusive access */
+    i2c_acquire(dev->i2c_dev);
+
+#if ENABLE_DEBUG
+    DEBUG("[Info] Access acquired !\n");
+#endif
+
+    /* Verify sensor ID */
+    uint8_t id;
+    i2c_read_reg(dev->i2c_dev, dev->addr,
+                 TSL2561_COMMAND_MODE | TSL2561_REGISTER_ID, &id);
+#if ENABLE_DEBUG
+    DEBUG("[Info] ID ? %d\n", id);
+#endif
+    if (id != TSL2561_ID ) {
+        DEBUG("[Error] not a TSL2561 sensor\n");
+        return -1;
```

Comment:
This is something different than stated in the API documentation, right?
`-1 if given I2C is not enabled in board config`